app/Libraries/Payments/XsollaPaymentProcessor.php

```diff
@@ -0,0 +1,118 @@
+<?php
+
+/**
+ *    Copyright 2015-2017 ppy Pty. Ltd.
+ *
+ *    This file is part of osu!web. osu!web is distributed with the hope of
+ *    attracting more community contributions to the core ecosystem of osu!.
+ *
+ *    osu!web is free software: you can redistribute it and/or modify
+ *    it under the terms of the Affero GNU General Public License version 3
+ *    as published by the Free Software Foundation.
+ *
+ *    osu!web is distributed WITHOUT ANY WARRANTY; without even the implied
+ *    warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ *    See the GNU Affero General Public License for more details.
+ *
+ *    You should have received a copy of the GNU Affero General Public License
+ *    along with osu!web.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+namespace App\Libraries\Payments;
+
+use App\Models\Store\Order;
+use Carbon\Carbon;
+
+// FIXME: rename?
+class XsollaPaymentProcessor extends PaymentProcessor
+{
+    const VALID_NOTIFICATION_TYPES = ['payment', 'refund', 'user_validation'];
+    const SKIP_NOTIFICATION_TYPES = ['user_search', 'user_validation'];
+
+    public function isSkipped()
+    {
+        return in_array($this->getNotificationType(), static::SKIP_NOTIFICATION_TYPES, true);
+    }
+
+    public function getOrderNumber()
+    {
+        return $this['transaction.external_id'];
+    }
+
+    public function getPaymentProvider()
+    {
+        return 'xsolla';
+    }
+
+    public function getPaymentTransactionId()
+    {
+        return $this['transaction.id'];
+    }
+
+    public function getPaymentAmount()
+    {
+        // TODO: less floaty
+        return (float) $this['purchase.checkout.amount'];
+    }
+
+    public function getPaymentDate()
+    {
+        return Carbon::parse($this['transaction.payment_date'])->setTimezone('UTC');
+    }
+
+    public function getNotificationType()
+    {
+        return $this['notification_type'];
+    }
+
+    public function validateTransaction()
```

Comment:
Should validate `allowedCheckoutTypes` otherwise someone can checkout by add supporter tag, open xsolla payment, open another tab to change cart content to contain physical items, and finish the xsolla payment.

![](https://s.myconan.net/2017-10/chrome_2017-10-17_18-31-01.png)